Implement delete-with-wildcards to remove entries from maps based on subkey values

BurntBrunch opened the pull request:
This PR implements the helper discussed in #2267.
Namely, the following is now a valid program:

```
tracepoint:syscalls:sys_enter_nanosleep
{
    @work[10, 10] = 50;
    @work[1, 10] = 10;
    @work[1, 20] = 100;
    @work[5, 20] = 100;
    @work[20, 20] = 100;

    delete_by_key_filter(@work, 1, 10); // deletes @work[10, 10] and @work[1, 10]
    delete_by_key_filter(@work, 0, 5); // deletes @work[5, 20]
}
```
Underneath, this calls bpf_map_for_each_elem() with a new static callback function. The callback receives as its context the key sub-value offset in bytes and the value to compare against. It currently only supports u64 comparisons.

It has been a bit a of a journey to get here:

1. bpftrace now emits debug info for procedure definitions
2. bpftrace now knows how to relocate BPF subprograms
3. bpftrace now knows how to relocate BTF func_info
4. bpftrace now passes BTF and func_info to bpf_prog_load calls

I'm happy to split this PR into smaller pieces too, if that's easier to review. Testing has been pretty limited, but I don't think I've broken anything with all this refactoring.

TODO:
  

- [x] Tests - open to suggestions. Not sure if codegen fixtures would help all that much but happy to do them.
- [x] Update documentation and changelog

Closes #2267.

viktormalik reviewed (commented):
This is a great amount of changes!

I reviewed everything but I'm not familiar with relocations enough to tell if they are correctly done. The rest looks good to me, just a couple of nits and questions.

What I'm concerned about is how this will have to change after #2322 is addressed. IIUC, once we have strong dependency on libbpf, we'll let it do relocations for us and large parts of the code proposed here will be removed. That said, I like the refactorings done in this PR (e.g. introduction of `BpfProgram`) but I'm not sure if makes sense to merge the rest if it'll go away soon.

viktormalik commented:
> * [ ]  Tests - open to suggestions. Not sure if codegen fixtures would help all that much but happy to do them

Well, the obvious basics - unit tests for semantic analysis of `delete_by_key_filter`, including correct and incorrect calls. Then, it'd be nice to have some runtime tests, it should be possible to use bpftool (such as [here](https://github.com/iovisor/bpftrace/blob/master/tests/runtime/probe#L52)) to check if the entries are properly deleted.

Codegen tests may not be that useful, but there's quite a lot of new code being generated, so I'd still add at least one test to prevent breaking it in future.

It'd be also nice to test the relocations somehow, but I have no clear idea how to do that.

BurntBrunch commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -38,10 +48,25 @@ CodegenLLVM::CodegenLLVM(Node *root, BPFtrace &bpftrace)
+                           0,
+                           StringRef(),
+                           DICompileUnit::DebugEmissionKind::LineTablesOnly);
+  debug_file_ = debug_.createFile("bpftrace", ".");
```
It doesn't _have to_ but I'll clean it up.

BurntBrunch commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -501,6 +526,42 @@ void CodegenLLVM::visit(Call &call)
+    auto key_value_deleter = accept(key_value_node);
+    auto *key_value = expr_;
+
+    // 16 bytes of context - 8 for key level and 8 for key value
```
Yup, will fix

BurntBrunch commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -501,6 +526,42 @@ void CodegenLLVM::visit(Call &call)
+    assert(key_level_node->is_literal && key_level_node->type.IsIntegerTy());
+    // Convert key level into key offset
+    int key_offset = 0;
+    int key_level_value = dynamic_cast<Integer *>(key_level_node)->n;
```
Sure, didn't know that existed.

BurntBrunch commented on src/ast/passes/codegen_llvm.cpp:
```diff
@@ -3050,6 +3254,60 @@ void CodegenLLVM::generateWatchpointSetupProbe(
+                                   i,
+                                   debug_file_,
+                                   0,
+                                   (DIType *)types[i],
```
It is statically known to be safe though. The types array is of an abstract type (`Metadata*`) but we populate it with `DIType *` ourselves. There's no need for dynamic type checks here imo.

viktormalik replied:
I agree it is safe, but I'd still put it there to prevent future problems if someone tries to add something that is not a `DIType*`.

BurntBrunch commented on src/bpfprogram.cpp:
```diff
@@ -0,0 +1,293 @@
+void BpfProgram::relocateMaps()
+{
+  struct bpf_insn *insns = reinterpret_cast<struct bpf_insn *>(code_.data());
+  for (uintptr_t i = 0; i < code_.size(); ++i)
```
No, you're right. I'll run this whole thing through ASAN, I might have missed something else. (We could make asan the default for debug builds?)

BurntBrunch replied:
No, you're right. I'll run this whole thing through ASAN, I might have missed something else. (We could make asan the default for debug builds?)

viktormalik replied:
It's run in the CI so we shouldn't miss anything. But I'm not against making it default for debug builds.

BurntBrunch commented:
Updates in this reroll:
1. User-facing documentation
2. Codegen tests updated for both debug info and delete_filter
3. Semantic analyzer and runtime tests
4. clang-format every commit
5. Addressed most of the feedback (didn't add that dynamic_cast though)

BurntBrunch commented:
@viktormalik what's the best way to reproduce the CI environment? The Vagrantfile seems out of date (a bunch of the images are not in vagrantcloud anymore).

Is there a convenient way to deal with these diverse (LLVM 9-13) environments or do I have to wrestle docker myself?

viktormalik commented:
> @viktormalik what's the best way to reproduce the CI environment? The Vagrantfile seems out of date (a bunch of the images are not in vagrantcloud anymore).
> 
> Is there a convenient way to deal with these diverse (LLVM 9-13) environments or do I have to wrestle docker myself?

Yes, the best way is to go with Docker. But you can basically reproduce the commands from `.github/workflows/ci.yml` to build and run an image. Some of the Vagrant images should be working, though, but they don't always contain what you need.

BurntBrunch commented:
Okay, I've hit a bit of a roadblock with the BTF support in this PR. Namely, LLVM only learns about BTF in version 10.

I can definitely gate the runtime code on the LLVM_VERSION but I can't really gate the grammar - I'd have to fail that in the semantic analyzer. 

I'd also have to make BTF program data optional in the whole pipeline, which is a shame - I really wanted to eventually add proper line info for that sweet inline source code when verification fails. Besides, as we move towards using libbpf for maps and object loading, it needs BTF for map definitions. 

I'd really like to bump the minimum to LLVM 10 as part of this stack.

Thoughts?

viktormalik commented:
> I'd really like to bump the minimum to LLVM 10 as part of this stack.
> 
> Thoughts?

Even though we have the minimum LLVM set to 6 in CMake, we run CI for 9+ only, so we basically don't support anything below. So dropping 9 and going with 10+ doesn't seem that big of a change and BTF support is clearly a strong enough reason to do so.

BurntBrunch commented:
> we run CI for 9+ only, so we basically don't support anything below.

Yes, I also want to clean that up in the code, I don't we can claim support for things we don't even build. 

> So dropping 9 and going with 10+ doesn't seem that big of a change and BTF support is clearly a strong enough reason to do so.

I'll send it as a separate PR :+1:

BurntBrunch commented:
I knew the timeline didn't make sense, this was an ORCv1 issue after all.
LLVM min version doesn't need changing from 9, though we should probably document that somewhere. (BTF was [added in LLVM 8](https://github.com/llvm/llvm-project/commit/7b410ac352bea299441a34f3b899e728bae460a7)). There's still a fair number of code branches we could clean up with 9 being the min.

ajor commented:
Definitely a good feature, but I'm not too keen on the proposed syntax. What about something like this as an alternative:

```
delete(@work[_, 10]); // deletes @work[10, 10] and @work[1, 10]
delete(@work[5, _]); // deletes @work[5, 20]
```
(or using asterisks)

We can have the semantic analyser reject things of this form that specify more than a single "key level" to filter on or that pass the result into functions other than `delete`, but it'd allow the syntax to be extended in the future to support more cases.

BurntBrunch commented:
@ajor thanks for the input! I can do a wildcard syntax, though there isn't a precedent for that - do we really want `_` to be a token or should we use `*` instead?

viktormalik commented:
> do we really want `_` to be a token or should we use `*` instead?

I'd go with `*`, that reflects better the purpose of the syntax (matching any value for that key level).

BurntBrunch commented:
This was a bit tricky to implement but I think I did with the least disruption to the grammar.

The idea is that map vargs are now special-cased (so different from array vargs) and they can accept "*" (token MUL) as a valid element. We introduce `ast::MapWildcard` type and `ast::Map::wildcards` field to track usage of wildcards in the expression. We then go and ignore any `Map` nodes with wildcards for the purposes of type deduction. Using `Map` nodes with wildcards anywhere but in `delete()` calls is an error and in `delete()` calls they get validated for completeness.

The docs and tests have more examples but overall I like this syntax.

viktormalik commented:
I did another round of review, cleaned up, and rebased onto master. Should be now ready for reviewing.

fbs commented:
Should the proposal about functions influence the design of this in any way?

viktormalik commented:
> Should the proposal about functions influence the design of this in any way?

It is possible that generic functions will require more types of relocations and a better solution will be required in future (possibly letting libbpf do the relocations). But I think that this is a good start and we'll have something to test against if we consider some bigger refactoring.

danobi commented:
Viktor is currently integrating this PR into master
